### PR TITLE
Initial support for Wii Hagi emulator

### PR DESCRIPTION
**Pull request by liamwhite**
This allows Galaxy and Sunshine to render and not just be completely black.

Unresolved issues:
- Mario's body and many scene elements do not appear yet (see Ryujinx/Ryujinx#1608)
- Both games absolutely chug and seem to be much slower than they really should be

010049900f546002_2022-03-13_22-58-31-638
010049900f546003_2022-03-13_22-28-31-482

### CODE REVIEW

**Comment by FernandoS27** [comment]
funny I have this implemented in yfc but the heads/gloves of the characters do render there. The rest of the character uses an indirect addressing const bank mode:

```
        /*00a4*/                  LEA.HI R4, R4, c[0x1][0x0], RZ, 0x1c ;
        /*00a8*/                   BFI R5, R5, 0x404, RZ ;
        /*00b0*/                   BFI R13, R1, 0x1010, R3 ;
        /*00b8*/                   LEA.HI R3, R0, c[0x1][0x0], RZ, 0x1c ;
        /*00c8*/                   LDC.64.IS R0, c[0x0][R13] ;
```

I have not fully REed that instruction though and how each mode works. Sadly I cannot look at the PR you linked as dealing with those stuffs bring us problems.

The rendering I called in my RE Direct3DDraw as it uses the exact same primitives as Direct3D even though it's the Vulkan driver that's behind this game. (really crazy and interesting driver so far). My guess is that both driver share it.

Now the question on is where did you get this registers as they are not in deko3d or nouveau's.

**Comment by liamwhite** [comment]
Hi,

I discovered these registers experimentally:
- I used a [modified yuzu](https://github.com/liamwhite/yuzu/tree/debug) that supports gdbstub;
- I [symbolicated the game](https://github.com/liamwhite/nx2elf/) and dumped info about its render calls; and
- I correlated the indirect draw calls with the register 0x5F9 write

This seems to match up with the [gf100 model in nouveau](https://github.com/envytools/envytools/blob/master/rnndb/graph/gf100_3d.xml#L1008-L1010).

This was not enough so then I also traced the render calls and found that they were rendering in points mode, so I also looked through the register assignments and found that there was an [override register](https://github.com/envytools/envytools/blob/master/rnndb/graph/gf100_3d.xml#L1095-L1107) that takes precedence over the normal draw register (this seems to be used to get around a macro function setting it to something undesirable)

**Comment by FernandoS27** [comment]
Thank you very much. You have my approval. Now there's a bug there, the head and gloves of mario should show up. let me see if I can find where your issue could be.

**Comment by liamwhite** [comment]
Sure, you can tag it for EA. The help is appreciated, it's been a long process discovering and fixing the issue.

**Comment by asLody** [comment]
@liamwhite You ignore that the dirty flag of IndexBuffer should be refreshed first.
```c++
dirty.flags[VideoCommon::Dirty::IndexBuffer] = true;
```

File: src/video_core/engines/maxwell_3d.cpp [hunk 1]
```diff
@@ -208,6 +208,10 @@ void Maxwell3D::ProcessMethodCall(u32 method, u32 argument, u32 nonshadow_argume
         return ProcessCBBind(4);
     case MAXWELL3D_REG_INDEX(draw.vertex_end_gl):
         return DrawArrays();
+    case MAXWELL3D_REG_INDEX(small_index):
+        regs.index_array.count = regs.small_index.count;
+        regs.index_array.first = regs.small_index.first;
+        return DrawArrays();
```
Comment (FernandoS27) [inline src/video_core/engines/maxwell_3d.cpp:214]: as AsLody said you must mark Indexbuffers as Dirty.

```
dirty.flags[VideoCommon::Dirty::IndexBuffer] = true;
```

**Comment by liamwhite** [comment]
Some more notes:
- I think there is another register that controls whether the topology overrides are enabled, but I don't know which one it is, so I have resorted to the kludge above, which enables overrides only after the override register is first set. This fixes other games that don't use it.
- The performance issues are caused by the emulator copying enormous vertex array buffers around without regard to their contents; for playtesting purposes it can be bandaged with this patch which _only_ works in these two games (Sunshine and Galaxy) and breaks everything else:
  ```diff
   diff --git a/src/video_core/buffer_cache/buffer_cache.h b/src/video_core/buffer_cache/buffer_cache.h
   index fa26eb8b0..cde5d42ca 100644
   --- a/src/video_core/buffer_cache/buffer_cache.h
   +++ b/src/video_core/buffer_cache/buffer_cache.h
   @@ -1282,10 +1282,20 @@ void BufferCache<P>::UpdateVertexBuffer(u32 index) {
        if (!maxwell3d.dirty.flags[Dirty::VertexBuffer0 + index]) {
            return;
        }
   +    u32 first_vertex = 0;
   +    u32 num_vertices = 0;
   +    if (maxwell3d.regs.index_array.count) {
   +        // optimistically assume that the number of
   +        // vertices is less than the index array count
   +        num_vertices = maxwell3d.regs.index_array.first + maxwell3d.regs.index_array.count;
   +    } else {
   +        first_vertex = maxwell3d.regs.vertex_buffer.first;
   +        num_vertices = maxwell3d.regs.vertex_buffer.count;
   +    }
        const auto& array = maxwell3d.regs.vertex_array[index];
   -    const auto& limit = maxwell3d.regs.vertex_array_limit[index];
   -    const GPUVAddr gpu_addr_begin = array.StartAddress();
   -    const GPUVAddr gpu_addr_end = limit.LimitAddress() + 1;
   +    const u32 stride = maxwell3d.regs.vertex_array[index].stride;
   +    const GPUVAddr gpu_addr_begin = array.StartAddress() + first_vertex * stride;
   +    const GPUVAddr gpu_addr_end = gpu_addr_begin + num_vertices * stride + 1;
        const std::optional<VAddr> cpu_addr = gpu_memory.GpuToCpuAddress(gpu_addr_begin);
        const u32 address_size = static_cast<u32>(gpu_addr_end - gpu_addr_begin);
        const u32 size = address_size; // TODO: Analyze stride and number of vertices
   ```

**Comment by FernandoS27** [comment]
@liamwhite This is not our issue with performance.

image

**Comment by liamwhite** [comment]
Yes, profiler revealed that the emulator is submitting very large buffer copy requests as part of the command queue. Maybe the issue doesn't manifest on other cards, but on my AMD card I find that I am getting 25 fps in Galaxy without the patch and 60 fps with it; and 7 fps in Sunshine without it and 30 with it.

I've sent you a friend request, will follow up there.

**Comment by liamwhite** [comment]
Okay, some investigation revealed that the issue with FMVs and with some scene elements not appearing is due to my GPU not supporting VK_FORMAT_R16G16B16_SSCALED, which the game uses for drawing most scene elements. (Support for 6-byte vertex buffers is generally found in newer hardware.) Fixing this is out of scope for this PR but it can be worked around like this:
```diff
diff --git a/src/video_core/renderer_vulkan/maxwell_to_vk.cpp b/src/video_core/renderer_vulkan/maxwell_to_vk.cpp
index 1c136c410..68f5999c4 100644
--- a/src/video_core/renderer_vulkan/maxwell_to_vk.cpp
+++ b/src/video_core/renderer_vulkan/maxwell_to_vk.cpp
@@ -404,7 +404,7 @@ VkFormat VertexFormat(Maxwell::VertexAttribute::Type type, Maxwell::VertexAttrib
         case Maxwell::VertexAttribute::Size::Size_16_16:
             return VK_FORMAT_R16G16_SSCALED;
         case Maxwell::VertexAttribute::Size::Size_16_16_16:
-            return VK_FORMAT_R16G16B16_SSCALED;
+            return VK_FORMAT_R16G16B16A16_SSCALED;
         case Maxwell::VertexAttribute::Size::Size_16_16_16_16:
             return VK_FORMAT_R16G16B16A16_SSCALED;
         case Maxwell::VertexAttribute::Size::Size_10_10_10_2:
```
I have updated the PR description to reflect the current state of these changes.

**Comment by FernandoS27** [comment]
> ```diff
> VK_FORMAT_R16G16B16_SSCALED
> ```

uhm this replacement is only possible if the stride is a multiple of the new format's size. i would suggest managing on the device.

File: src/video_core/engines/maxwell_3d.cpp [hunk 2]
```diff
@@ -370,8 +370,29 @@ void Maxwell3D::CallMethodFromMME(u32 method, u32 method_argument) {
 }
 
 void Maxwell3D::ProcessTopologyOverride() {
+    using PrimitiveTopology = Maxwell3D::Regs::PrimitiveTopology;
+    using PrimitiveTopologyOverride = Maxwell3D::Regs::PrimitiveTopologyOverride;
+
+    PrimitiveTopology topology{};
+
+    switch (regs.topology_override) {
+    case PrimitiveTopologyOverride::None:
```
Comment (FernandoS27) [inline src/video_core/engines/maxwell_3d.cpp:379]: In this one it should use the original register.

**Review by FernandoS27 — APPROVED** [review]

**Pull request merged**